File: backend/app/schemas/notification.py
"""Pydantic v2 schemas for the notification framework."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Template schemas
# ---------------------------------------------------------------------------


class NotificationTemplateOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    template_key: str
    channel: str
    subject: Optional[str] = None
    body: str
    is_active: bool
    dlt_template_id: Optional[str] = None
    dlt_entity_id: Optional[str] = None
    approved_variables: Optional[list[str]] = None
    created_at: datetime
    updated_at: datetime


class NotificationTemplateUpdate(BaseModel):
    """Only editable fields — template_key and dlt_template_id are immutable."""

    subject: Optional[str] = Field(None, max_length=255)
    body: Optional[str] = Field(None, min_length=1)
    is_active: Optional[bool] = None
    approved_variables: Optional[list[str]] = None


# ---------------------------------------------------------------------------
# Log schemas
# ---------------------------------------------------------------------------


class NotificationLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    template_key: str
    channel: str
    recipient_email: Optional[str] = None
    recipient_mobile: Optional[str] = None
    entity_type: Optional[str] = None
    entity_id: Optional[str] = None
    subscription_id: Optional[UUID] = None
    days_offset: Optional[int] = None
    provider_name: Optional[str] = None
    provider_message_id: Optional[str] = None
    status: str
    error_message: Optional[str] = None
    created_at: datetime
    sent_at: Optional[datetime] = None


class NotificationLogListResponse(BaseModel):
    items: list[NotificationLogOut]
    total: int
    page: int
    page_size: int
    total_pages: int


# ---------------------------------------------------------------------------
# Preference schemas
# ---------------------------------------------------------------------------


class NotificationPreferenceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    customer_id: UUID
    welcome_sms_enabled: bool
    welcome_email_enabled: bool
    renewal_sms_enabled: bool
    renewal_email_enabled: bool
    invoice_email_enabled: bool
    payment_email_enabled: bool
    otp_sms_enabled: bool
    otp_email_enabled: bool
    created_at: datetime
    updated_at: datetime


class NotificationPreferenceUpdate(BaseModel):
    welcome_sms_enabled: Optional[bool] = None
    welcome_email_enabled: Optional[bool] = None
    renewal_sms_enabled: Optional[bool] = None
    renewal_email_enabled: Optional[bool] = None
    invoice_email_enabled: Optional[bool] = None
    payment_email_enabled: Optional[bool] = None
    otp_sms_enabled: Optional[bool] = None
    otp_email_enabled: Optional[bool] = None


# ---------------------------------------------------------------------------
# Test-send schemas
# ---------------------------------------------------------------------------


class TestEmailRequest(BaseModel):
    to_email: str = Field(..., min_length=1)
    subject: str = Field(default="Test Email from True Data Broadband")
    body: str = Field(default="This is a test email to verify SMTP configuration.")


class TestSmsRequest(BaseModel):
    to_mobile: str = Field(..., min_length=10, max_length=15)
    message: str = Field(default="Test SMS from True Data Broadband.")


class TestSendResponse(BaseModel):
    success: bool
    message: str
    provider_response: Optional[Any] = None

File: frontend/src/services/notification.ts
import { api } from "./api";
import type {
  NotificationLogPage,
  NotificationPreference,
  NotificationTemplate,
  NotificationTemplateUpdate,
} from "@/types/notification";

const BASE = "/notifications";

// ── Templates ─────────────────────────────────────────────────────────────

export async function listTemplates(): Promise<NotificationTemplate[]> {
  const { data } = await api.get<NotificationTemplate[]>(`${BASE}/templates`);
  return data;
}

export async function updateTemplate(
  id: string,
  payload: NotificationTemplateUpdate,
): Promise<NotificationTemplate> {
  const { data } = await api.put<NotificationTemplate>(
    `${BASE}/templates/${id}`,
    payload,
  );
  return data;
}

// ── Logs ──────────────────────────────────────────────────────────────────

export interface LogFilter {
  page?: number;
  page_size?: number;
  template_key?: string;
  channel?: string;
  status?: string;
  date_from?: string;
  date_to?: string;
}

export async function listLogs(params: LogFilter = {}): Promise<NotificationLogPage> {
  const { data } = await api.get<NotificationLogPage>(`${BASE}/logs`, { params });
  return data;
}

// ── Preferences ───────────────────────────────────────────────────────────

export async function getPreferences(customerId: string): Promise<NotificationPreference> {
  const { data } = await api.get<NotificationPreference>(
    `${BASE}/preferences/${customerId}`,
  );
  return data;
}

export async function updatePreferences(
  customerId: string,
  payload: Partial<NotificationPreference>,
): Promise<NotificationPreference> {
  const { data } = await api.put<NotificationPreference>(
    `${BASE}/preferences/${customerId}`,
    payload,
  );
  return data;
}

// ── Test send ─────────────────────────────────────────────────────────────

export async function testEmail(params: {
  to_email: string;
  subject?: string;
  body?: string;
}): Promise<{ success: boolean; message: string }> {
  const { data } = await api.post(`${BASE}/test-email`, params);
  return data;
}

export async function testSms(params: {
  to_mobile: string;
  message?: string;
}): Promise<{ success: boolean; message: string }> {
  const { data } = await api.post(`${BASE}/test-sms`, params);
  return data;
}

export const notificationService = {
  listTemplates,
  updateTemplate,
  listLogs,
  getPreferences,
  updatePreferences,
  testEmail,
  testSms,
};
